File: frontend/app/types/transaction.ts
/**
 * Transaction and financial types for the Learning Management System.
 */

export enum TransactionStatus {
    PENDING = "pending",
    COMPLETED = "completed",
    FAILED = "failed",
    CANCELLED = "cancelled",
    REFUNDED = "refunded"
}

export enum PaymentMethod {
    STRIPE = "stripe",
    PAYPAL = "paypal",
    BANK_TRANSFER = "bank_transfer",
    FREE = "free"
}

export interface Transaction {
    id: number;
    transaction_id: string;
    user_id: number;
    course_id?: number;
    amount: number;
    currency: string;
    status: TransactionStatus;
    payment_method: PaymentMethod;
    gateway_transaction_id?: string;
    refund_amount: number;
    refund_reason?: string;
    refunded_at?: string;
    description?: string;
    notes?: string;
    created_at: string;
    updated_at: string;
    completed_at?: string;
    net_amount: number;
}

export interface TransactionListResponse {
    transactions: Transaction[];
    total: number;
    page: number;
    per_page: number;
    total_pages: number;
}

export interface TransactionFilter {
    user_id?: number;
    course_id?: number;
    status?: TransactionStatus;
    payment_method?: PaymentMethod;
    start_date?: string;
    end_date?: string;
    min_amount?: number;
    max_amount?: number;
}

export interface PaymentIntentCreate {
    course_id: number;
    payment_method?: PaymentMethod;
    return_url?: string;
}

export interface PaymentIntentResponse {
    client_secret: string;
    transaction_id: string;
    amount: number;
    currency: string;
}

export interface RefundCreate {
    refund_amount: number;
    refund_reason: string;
}

export interface InstructorPayout {
    id: number;
    payout_id: string;
    instructor_id: number;
    amount: number;
    currency: string;
    commission_rate: number;
    period_start: string;
    period_end: string;
    status: TransactionStatus;
    payment_method?: string;
    external_payout_id?: string;
    created_at: string;
    updated_at: string;
    processed_at?: string;
}

export interface InstructorPayoutListResponse {
    payouts: InstructorPayout[];
    total: number;
    page: number;
    per_page: number;
    total_pages: number;
}

export interface InstructorEarnings {
    instructor_id: number;
    total_earnings: number;
    pending_earnings: number;
    paid_earnings: number;
    total_sales: number;
    commission_rate: number;
    period_start: string;
    period_end: string;
}

export interface RevenueOverview {
    period_start: string;
    period_end: string;
    total_transactions: number;
    gross_revenue: number;
    net_revenue: number;
    total_refunds: number;
    platform_revenue: number;
    total_instructor_payouts: number;
    avg_transaction_value: number;
    unique_customers: number;
    customer_ltv: number;
    refund_rate: number;
}

export interface RevenueTrend {
    period: string;
    transaction_count: number;
    gross_revenue: number;
    net_revenue: number;
    total_refunds: number;
}

export interface CourseRevenueAnalysis {
    course_id: number;
    course_title: string;
    course_price: number;
    instructor_name: string;
    sales_count: number;
    gross_revenue: number;
    net_revenue: number;
    total_refunds: number;
    refund_rate: number;
}

export interface InstructorRevenueAnalysis {
    instructor_id: number;
    instructor_name: string;
    course_count: number;
    sales_count: number;
    gross_revenue: number;
    net_revenue: number;
    total_refunds: number;
    estimated_commission: number;
    avg_revenue_per_course: number;
}

export interface PaymentMethodAnalysis {
    payment_method: string;
    transaction_count: number;
    total_revenue: number;
    avg_transaction_value: number;
    percentage_of_transactions: number;
    percentage_of_revenue: number;
}

export interface RefundAnalysis {
    period_start: string;
    period_end: string;
    total_transactions: number;
    total_refunded_transactions: number;
    refund_rate: number;
    total_refund_amount: number;
    refund_percentage_of_revenue: number;
    avg_refund_amount: number;
    refund_reasons: Record<string, { count: number; total_amount: number }>;
    top_refunded_courses: Array<{
        course_id: number;
        course_title: string;
        refund_count: number;
        refund_amount: number;
    }>;
}

export interface FinancialDashboardData {
    period_start: string;
    period_end: string;
    revenue_overview: RevenueOverview;
    revenue_trends: RevenueTrend[];
    top_courses: CourseRevenueAnalysis[];
    top_instructors: InstructorRevenueAnalysis[];
    payment_methods: PaymentMethodAnalysis[];
    pending_payouts: {
        count: number;
        total_amount: number;
    };
}

File: backend/app/schemas/transaction.py
"""
Transaction and payment schemas for the Learning Management System.
"""

from pydantic import BaseModel, Field, validator
from typing import Optional, List
from datetime import datetime
from decimal import Decimal
from enum import Enum

from ..models.transaction import TransactionStatus, PaymentMethod


class TransactionBase(BaseModel):
    """Base transaction schema."""
    amount: Decimal = Field(..., gt=0, description="Transaction amount")
    currency: str = Field(default="USD", max_length=3, description="Currency code")
    payment_method: PaymentMethod = Field(..., description="Payment method used")
    description: Optional[str] = Field(None, max_length=500, description="Transaction description")
    notes: Optional[str] = Field(None, description="Additional notes")


class TransactionCreate(TransactionBase):
    """Schema for creating a new transaction."""
    course_id: Optional[int] = Field(None, description="Course ID for course purchases")
    gateway_transaction_id: Optional[str] = Field(None, max_length=200, description="External gateway transaction ID")


class TransactionUpdate(BaseModel):
    """Schema for updating a transaction."""
    status: Optional[TransactionStatus] = None
    gateway_transaction_id: Optional[str] = Field(None, max_length=200)
    gateway_response: Optional[str] = None
    description: Optional[str] = Field(None, max_length=500)
    notes: Optional[str] = None
    completed_at: Optional[datetime] = None


class RefundCreate(BaseModel):
    """Schema for creating a refund."""
    refund_amount: Decimal = Field(..., gt=0, description="Amount to refund")
    refund_reason: str = Field(..., max_length=500, description="Reason for refund")


class TransactionResponse(TransactionBase):
    """Schema for transaction response."""
    id: int
    transaction_id: str
    user_id: int
    course_id: Optional[int]
    status: TransactionStatus
    gateway_transaction_id: Optional[str]
    refund_amount: Decimal
    refund_reason: Optional[str]
    refunded_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime]
    net_amount: Decimal

    class Config:
        from_attributes = True


class TransactionListResponse(BaseModel):
    """Schema for transaction list response."""
    transactions: List[TransactionResponse]
    total: int
    page: int
    per_page: int
    total_pages: int


class TransactionFilter(BaseModel):
    """Schema for filtering transactions."""
    user_id: Optional[int] = None
    course_id: Optional[int] = None
    status: Optional[TransactionStatus] = None
    payment_method: Optional[PaymentMethod] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    min_amount: Optional[Decimal] = None
    max_amount: Optional[Decimal] = None


class PaymentIntentCreate(BaseModel):
    """Schema for creating a payment intent."""
    course_id: int = Field(..., description="Course ID to purchase")
    payment_method: PaymentMethod = Field(default=PaymentMethod.STRIPE, description="Payment method")
    return_url: Optional[str] = Field(None, description="Return URL after payment")


class PaymentIntentResponse(BaseModel):
    """Schema for payment intent response."""
    client_secret: str = Field(..., description="Client secret for frontend")
    transaction_id: str = Field(..., description="Internal transaction ID")
    amount: Decimal = Field(..., description="Payment amount")
    currency: str = Field(..., description="Currency code")


class WebhookEvent(BaseModel):
    """Schema for webhook events."""
    event_type: str = Field(..., description="Type of webhook event")
    event_id: str = Field(..., description="Unique event ID")
    data: dict = Field(..., description="Event data")
    created: datetime = Field(..., description="Event creation time")


class InstructorPayoutBase(BaseModel):
    """Base instructor payout schema."""
    amount: Decimal = Field(..., gt=0, description="Payout amount")
    currency: str = Field(default="USD", max_length=3, description="Currency code")
    commission_rate: Decimal = Field(..., ge=0, le=100, description="Commission rate percentage")
    period_start: datetime = Field(..., description="Payout period start")
    period_end: datetime = Field(..., description="Payout period end")
    payment_method: Optional[str] = Field(None, max_length=50, description="Payout method")


class InstructorPayoutCreate(InstructorPayoutBase):
    """Schema for creating instructor payout."""
    instructor_id: int = Field(..., description="Instructor user ID")


class InstructorPayoutUpdate(BaseModel):
    """Schema for updating instructor payout."""
    status: Optional[TransactionStatus] = None
    external_payout_id: Optional[str] = Field(None, max_length=200)
    payout_details: Optional[str] = None
    processed_at: Optional[datetime] = None


class InstructorPayoutResponse(InstructorPayoutBase):
    """Schema for instructor payout response."""
    id: int
    payout_id: str
    instructor_id: int
    status: TransactionStatus
    external_payout_id: Optional[str]
    created_at: datetime
    updated_at: datetime
    processed_at: Optional[datetime]

    class Config:
        from_attributes = True


class InstructorPayoutListResponse(BaseModel):
    """Schema for instructor payout list response."""
    payouts: List[InstructorPayoutResponse]
    total: int
    page: int
    per_page: int
    total_pages: int


class InstructorEarnings(BaseModel):
    """Schema for instructor earnings summary."""
    instructor_id: int
    total_earnings: Decimal
    pending_earnings: Decimal
    paid_earnings: Decimal
    total_sales: int
    commission_rate: Decimal
    period_start: datetime
    period_end: datetime